experimental/HeapSparseLA/test/runtests.jl

```diff
@@ -0,0 +1,104 @@
+@testset "heap based sparse linear algebra" begin
+  n = Oscar.HeapNode(1, ZZ(5))
+  n.left = Oscar.HeapNode(2, ZZ(7))
+  n.right = Oscar.HeapNode(2, ZZ(-3))
+  n.right.right = Oscar.HeapNode(3, ZZ(8))
+  v = Oscar.HeapSRow(ZZ, n)
+  w = Oscar.HeapSRow(ZZ, [1 => ZZ(5), 2 => ZZ(4), 3 => ZZ(11), 3=>ZZ(-3)])
+  @test !isdefined(v, :index_cache)
+  @test v[3] == 8
+  @test isdefined(v, :index_cache) && v.index_cache[3] == 8
+  @test Oscar.tree_size(v.top) == 4
+  @test v[2] == 4
+  @test v.index_cache[2] == 4
+  @test Oscar.tree_size(v.top) == 3
+
+  u = v + w
+  @test u == 2*v
+end
+
+@testset "Gauss algorithm on different matrix types" begin
+  s = 4
+  m = 10*s
+  n = 20*s
+
+  A = Oscar.HeapSMat(ZZ, 0, n)
+  for i in 1:m
+    l = [(rand(1:20*s), ZZ(rand(-10:10))) for i in 1:s];
+    v = Oscar.HeapSRow(ZZ, l)
+    push!(A, v)
+  end
+  AA = Oscar.sparse_matrix(A)
+
+  B0, S0, T0 = Oscar.upper_triangular_form!(copy(A))
+  B1, S1, T1 = Oscar.upper_triangular_form!(copy(AA))
+
+  @test S0*A == B0
+  @test T0*B0 == A
+
+  S1 = Oscar.dense_matrix(Oscar.HeapSMat(S1))
+  B1 = Oscar.dense_matrix(Oscar.HeapSMat(B1))
+  T1 = Oscar.dense_matrix(Oscar.HeapSMat(T1))
+  A1 = Oscar.dense_matrix(A)
+  @test S1*A1 == B1
+  @test T1*B1 == A1
+
+  @test Oscar.HeapSMat(B1) == B0
+  @test Oscar.HeapSMat(T1) == T0
+  @test Oscar.HeapSMat(S1) == S0
+end
+
+@testset "Gauss algorithm on different matrix types over univariate polynomial rings" begin
+  s = 1
+  m = 10*s
+  n = 20*s
+
+  P, (x,) = polynomial_ring(QQ, [:x])
+  QQt, t = QQ[:t]
+  A = Oscar.HeapSMat(QQt, 0, n)
+  for i in 1:m
+    l = [(rand(1:20*s), evaluate(rand(P, 1:10, 1:10, 1:10), [t])) for i in 1:s];
+    v = Oscar.HeapSRow(QQt, l)
+    push!(A, v)
+  end
+  AA = Oscar.sparse_matrix(A)
+
+  B0, S0, T0 = Oscar.upper_triangular_form!(copy(A))
+  B1, S1, T1 = Oscar.upper_triangular_form!(copy(AA))
+
+  @test S0*A == B0
+  @test T0*B0 == A
+
+  S1 = Oscar.dense_matrix(Oscar.HeapSMat(S1))
+  B1 = Oscar.dense_matrix(Oscar.HeapSMat(B1))
+  T1 = Oscar.dense_matrix(Oscar.HeapSMat(T1))
+  A1 = Oscar.dense_matrix(A)
+  @test S1*A1 == B1
+  @test T1*B1 == A1
+
+  @test Oscar.HeapSMat(B1) == B0
+  @test Oscar.HeapSMat(T1) == T0
+  @test Oscar.HeapSMat(S1) == S0
+end
+
+@testset "heap-dict vectors" begin
+  k = 100000
+  i = unique!([rand(1:10*k) for i in 1:k])
+  v = [ZZ(rand(1:10))^10000 for i in 1:length(i)]
+  a = Oscar.HeapDictSRow(ZZ, i, v)
+  aa = sparse_row(a)
+  i = unique!([rand(1:10*k) for i in 1:k])
+  v = [ZZ(rand(1:10))^10000 for i in 1:length(i)]
+  b = Oscar.HeapDictSRow(ZZ, i, v)
+  bb = sparse_row(b)
+
+  # For comparison of timings do 
+  # julia> c = copy(a); @time w = add!(c, b);
+  #   0.042619 seconds (27.40 k allocations: 41.765 MiB, 17.17% gc time)
+  #
+  # julia> cc = copy(aa); @time cc += bb;
+  #   0.226098 seconds (192.12 k allocations: 510.190 MiB, 5.40% gc time)
+
```

Comment:
Looks like we are slowly getting somewhere. But this only works for `RingElem`s which are big in the sense that copying them around in memory is expensive. For small elements it seems that the original `SRow` is still winning. In that case the rehashing while merging the dictionaries is the expensive part.

Reply:
@joschmitt There are other implementations that use other finite fields with the same approach, e.g. https://link.springer.com/article/10.1007/s41468-021-00071-5. I have no idea how the heap based approach behaves in char zero. But I'd try that. Also, the cohomology computation where this approach is beneficial is quite specific (clique complexes), so it may not generalize. @HechtiDerLachs I'll look at your code soon and will play around with the underlying heap.